Features/lease preference used #157669475

akegan opened the pull request:
https://www.pivotaltracker.com/story/show/157669475

akegan committed:
[#157669475] Show confirmed preferences in dropdown

akegan committed:
[#157669475] Add 'None' as an option

akegan committed:
[#157669475] Wire up preferenceUsed to backend

akegan committed:
Merge branch 'master' into features/lease-preference-used-#157669475

akegan committed:
[#157669475] Save and load preference used, no preference used

fddayan commented on app/javascript/components/supplemental_application/sections/LeaseInformatonInputs.js:
```diff
@@ -37,6 +37,19 @@ const LeaseInformationInputs = ({ formApi, store }) => {
   const availableUnitsOptions = formUtils.toOptions(map(availableUnits, pluck('id', 'unit_number')))
   const totalMonthlyRent = getTotalMonthlyRent(formApi.values)
+  // FIXME: will this dynamically update when the preferences are updated?
+  const preferences = store.application.preferences
```
application can be unpacked in line 37

akegan commented on app/javascript/components/supplemental_application/SupplementalApplicationPage.js:
```diff
@@ -71,6 +71,14 @@ class SupplementalApplicationPage extends React.Component {
     synchedApplication.preferences = cloneDeep(persistedApplication.preferences)
 
+    // Handle lease preference used logic.
+    if (synchedApplication.lease.preference_used === 'no_preference_used') {
```
try using null as the id instead of 'no_preference_used'

akegan replied:
doesn't work

akegan committed:
Merge branch 'master' into features/lease-preference-used-#157669475

akegan commented on app/javascript/components/supplemental_application/sections/LeaseInformatonInputs.js:
```diff
@@ -52,7 +65,7 @@ const LeaseInformationInputs = ({ formApi, store }) => {
         <CustomFormGrid.Select
           label='Preference Used'
-          field='preference_used' options={[]} placeholder='Select One' />
+          field='preference_used' options={confirmedPreferenceOptions} placeholder='Select One' />
```
look into a hidden field

akegan committed:
[#157669475] Send postLotteryValidation with shortform

chrisdolendo committed:
[#157669475] - Implement toggle of noPreferenceUsed on Lease based on change on preferenceUsed

chrisdolendo committed:
[#157669475] - remove string value of no_preference_used for None selection for preferenceUsed, update ui logic

chrisdolendo commented on app/javascript/components/supplemental_application/sections/LeaseInformatonInputs.js:
```diff
@@ -52,7 +68,8 @@ const LeaseInformationInputs = ({ formApi, store }) => {
         <CustomFormGrid.Select
           label='Preference Used'
-          field='preference_used' options={[]} placeholder='Select One' />
+          onChange={toggleNoPreferenceUsed(formApi)}
```
@akegan created an onchange function so that the value assignment for `noPreferenceUsed` stays within this file :)

chrisdolendo commented on app/javascript/components/supplemental_application/sections/LeaseInformatonInputs.js:
```diff
@@ -33,10 +33,26 @@ const getTotalMonthlyRent = (values) => {
+  // FIXME: will this dynamically update when the preferences are updated?
+  const preferences = store.application.preferences
+  const confirmedPreferences = filter(preferences, { 'post_lottery_validation': 'Confirmed' })
+  const confirmedPreferenceOptions = formUtils.toOptions(map([{'id': null, 'preference_name': 'None'}, ...confirmedPreferences], pluck('id', 'preference_name')))
```
I just changed the `no_preferenced_used` string to `null` so that the value doesn't need to be changed to null on submit. I would vote to change the name of 'id' to 'value', but the functionality still works either way.

chrisdolendo commented on app/javascript/components/supplemental_application/SupplementalApplicationPage.js:
```diff
@@ -233,8 +233,8 @@ const getAnnualIncome = ({ monthlyIncome, annualIncome }) => {
   const applicationWithDefaults = cloneDeep(application)
-
   applicationWithDefaults.annual_income = getAnnualIncome({monthlyIncome: application.monthly_income, annualIncome: application.annual_income})
+  applicationWithDefaults.lease.no_preference_used = !application.lease.preference_used
```
I initially moved this to the `LeaseInformationInputs` component but kept it here so it stays centralized to this function.

chrisdolendo committed:
[#157669475] - Refactor lease service to simplify params

chrisdolendo commented on app/services/force/lease_service.rb:
```diff
@@ -30,28 +31,27 @@ def submit_lease(lease, application_id, primary_contact_id)
-                      Monthly_Parking_Rent__c: lease[:monthlyParkingRent],
-                      Total_Monthly_Rent_without_Parking__c: lease[:totalMonthlyRentWithoutParking],
-                      Monthly_Tenant_Contribution__c: lease[:monthlyTenantContribution])
+      create_lease_params = lease_base_params(lease, application_id).merge(Tenant__c: primary_contact_id, Lease_Status__c: 'Draft')
```
A little refactor here and on update so we can clearly see the difference in the params.

chrisdolendo committed:
[#157669475] - Simplified logic on assigning lease section default values on supp app"

chrisdolendo commented on app/javascript/components/supplemental_application/SupplementalApplicationPage.js:
```diff
@@ -233,9 +233,11 @@ const getAnnualIncome = ({ monthlyIncome, annualIncome }) => {
   applicationWithDefaults.annual_income = getAnnualIncome({monthlyIncome: application.monthly_income, annualIncome: application.annual_income})
-
+  // Logic in Lease Section in order to show 'Select One' placeholder on Preference Used if a selection was never made
+  if (applicationWithDefaults.lease && !applicationWithDefaults.lease.no_preference_used && applicationWithDefaults.lease.preference_used == null) {
```
The dropdown for `preferenceUsed` works such as:
If the value of `preference_used` is `null`, the dropdown will show "None".
If there is no `preference_used` key in the lease object, the dropdown will show the placeholder "Select One".

Therefore, I created this logic (@akegan this was based on our discussion) to delete the key, and know for certain that `preference_used` was an untouched field when first loading the page.

akegan commented on app/services/force/lease_service.rb:
```diff
@@ -31,32 +31,27 @@ def submit_lease(lease, application_id, primary_contact_id)
-                      No_Preference_Used__c: lease[:noPreferenceUsed],
-                      Total_Monthly_Rent_without_Parking__c: lease[:totalMonthlyRentWithoutParking],
-                      Monthly_Tenant_Contribution__c: lease[:monthlyTenantContribution])
+      create_lease_params = lease_base_params(lease, application_id).merge(Id: lease[:id])
```
we need the lease status in here too

akegan committed:
[#157669475] Remove console.logs

akegan committed:
[#157669475] Update snapshots

akegan commented:
I removed console logs and updated the snapshots. Looks good to me except for the comment above about including the lease status on update. 

I think that the FIXME about wondering how it will work with the preference update is still valid, and we won't really be able to know for sure until @callaghanc's PR gets merged. I'm up for creating a followup ticket or waiting til it's ready for QA and then rebasing on top. Either way.

chrisdolendo committed:
[#157669475] - Include lease status in lease updating

chrisdolendo commented:
@akegan added my update here, and handing back off to you. Let me know if you need to pair.

akegan committed:
Merge branch 'master' into features/lease-preference-used-#157669475

akegan committed:
[#157669475] Move where application is unpacked from store

callaghanc committed:
Merge branch 'master' into features/lease-preference-used-#157669475

akegan commented:
Update: currently the dropdown does not update based on status updates on the preferences being saved. I need to look into what is updating and what is not updating and possible pair with @fddayan tomorrow morning to discuss

akegan commented:
Blocked by this interaction pr: https://github.com/Exygy/sf-dahlia-lap/pull/104 

(That PR will fix the issue that the dropdown options are not updated when preferences are confirmed)

akegan committed:
Merge branch 'master' into features/lease-preference-used-#157669475

callaghanc committed:
Merge branch 'master' into features/lease-preference-used-#157669475

callaghanc commented:
@akegan We've just merged the confirmed prefs panel UX behavior PR, so this PR is now unblocked!

akegan commented:
Testing this now to see if it works as-is

akegan commented:
It is very close to working, but not quite. the available options in the dropdown only change if you change the selection, then the next time you click on it you get the updated options 😕. looking into it more now

akegan committed:
[#157669475] Update preference used dropdown on preference change

akegan committed:
[#157669475] Update snapshots

akegan committed:
Merge branch 'master' into features/lease-preference-used-#157669475

akegan commented:
This is ready for review. I'm working on adding tests, but I'm wondering if we should get this an initial dev review and then we can start on product acceptance while I work on figuring out how to add some unit tests

fddayan commented on app/javascript/components/supplemental_application/sections/LeaseInformationInputs.js:
```diff
@@ -52,7 +58,8 @@ const LeaseInformationInputs = ({ formApi, store }) => {
         <CustomFormGrid.Select
           label='Preference Used'
-          field='preference_used' options={[]} placeholder='Select One' />
+          onChange={toggleNoPreferenceUsed(formApi)}
```
What is the intention of this? 
`toggleNoPreferenceUsed` in line 36 should be returning a function that can be used as a the event handler for `onChange`. Any particular reason we are doing this?

chrisdolendo replied:
@fddayan I wrote this code so responding -- So you're saying it should be `onChange={toggleNoPreferenceUsed}` instead of `onChange={toggleNoPreferenceUsed(formApi)}`, I remember I tried using the former (`{toggleNoPreferenceUsed}`) but it wasn't working 100% correctly.

fddayan commented on app/javascript/components/supplemental_application/sections/LeaseInformationInputs.js:
```diff
@@ -33,10 +33,16 @@ const getTotalMonthlyRent = (values) => {
 }
 
+const toggleNoPreferenceUsed = (formApi) => {
+  if (formApi.touched.preference_used) { formApi.values.no_preference_used = !formApi.values.preference_used }
```
values should not be modified directly. Instead `formApi.setValue(field, value)` should be used.
react-form refs for our version: https://dazzling-keller-a8107a.netlify.com/#/

Reading from `formApi.value` is ok 👍

chrisdolendo replied:
@fddayan Good catch, I will change to use `setValue` instead.

chrisdolendo committed:
[#157669475] - Use .setValue on formApi to toggle noPreferenceUsed

fddayan reviewed (approved):
looks good 👍

chrisdolendo committed:
[#157669475]- Merged master

chrisdolendo committed:
[#157669475]- Updated toggling of no pref to let tests pass

chrisdolendo commented on app/javascript/components/supplemental_application/sections/LeaseInformationInputs.js:
```diff
@@ -52,7 +58,8 @@ const LeaseInformationInputs = ({ formApi, store }) => {
         <CustomFormGrid.Select
           label='Preference Used'
-          field='preference_used' options={[]} placeholder='Select One' />
+          onChange={(value) => toggleNoPreferenceUsed(formApi, value)}
```
@fddayan 😄
